### DISCUSSION TITLE
Horizontal scalability approach

### DISCUSSION THREAD

**Discussion by tommoor** [ueberdosis/hocuspocus#133] [General]
I was looking through the code for the Websocket extension trying to find how it deals with horizontal scalability and nothing really jumped out at me. I was wondering what are the teams thoughts on how to deal with this? For example:

- I several hocuspocus servers running connected over redis
- Each has a person connected and editing
- Is each server debouncing separately and causing a cascade of webhooks?

If so have you had any thoughts about how to handle this in a more efficient manner yet?

**Comment by hanspagel** [comment]
We’re waiting for a y-redis rewrite from Kevin, and will start tackling that then. The HA setup we aim for is multiple hocuspocus nodes in the front with a hocuspocs worker (e.g. for Webhooks) in the background.

**Reply by tommoor** [reply]
I've seen the worker language a couple of times. Wouldn't this need to know about edits across all nodes to be in control of websockets? It then becomes a bottleneck.

Is it just another server node without client connections but with extra extensions? Any pointers where this has been discussed?

**Reply by hanspagel** [reply]
Nodes could load the required document from Redis. If two nodes need the same document, both can subscribe to the same node through Redis. 

Let’s say you’re connected through a node in Asia and I’m through a node in South America, yes updates would need to travel around the world (through Redis or a Redis Cluster). 

BTW Y.js allows to apply updates without actually loading the document. This is the most performanent way to handle updates. We’ll add that feature to tiptap once we start tackling the node/worker concept.

And yes, the worker wouldn’t have any client connections. It would pick up content from Redis  to do "slow" tasks, like writing to Postgres or sending Webhooks. And yes, nodes and workers would have a different set of extensions then.

There is closed issue discussing scalability, there’s not much more info. 

Let me know if you need to know more or have other ideas. ✌️

**Reply by tommoor** [reply]
Okay, so it sounds like there is another process that subscribes to all redis updates and writes yjs docs to Postgres on a scheduled basis but this doesn't exist in the codebase at all yet.

As I understand it `y-redis` is a persistence layer rather than designed for horizontal event propagation. Perhaps the redis layer should be brought into hocuspocus rather than waiting on Kevin?

**Reply by hanspagel** [reply]
More like a queue taking all inactive documents out of Redis and persisting them (Postgres, Mongo, Webhooks …), yes!

And yes, you’re right. It wouldn’t hurt to start implementing a few of the described features before the y-redis rewrite is ready to use. It’s just that we’re taking things slow. ✌️ Let me know if you’d like to contribute financially or in code to speed things up, happy to jump on a call with you at some point: humans@tiptap.dev
